### PR TITLE
Refactored AutoGen integration + added examples folder

### PR DESCRIPTION
**Pull request by cpacker**
TODO

- [x] Make sure PM/coder example is working as intended

~~- [ ] Add TeachableAgent example (from https://github.com/microsoft/autogen/blob/main/test/agentchat/chat_with_teachable_agent.py / https://github.com/microsoft/autogen/blob/main/notebook/agentchat_teachability.ipynb)~~

### CODE REVIEW

**Comment by cpacker** [comment]
@vivi Running with this persona/user setting + initial user_proxy message:

```python
...

    coder = create_autogen_memgpt_agent(
        "MemGPT_coder",
        persona_description="I am a 10x engineer, trained in Python. I was the first engineer at Uber (which I make sure to tell everyone I work with).",
        user_description="You are participating in a group chat with a user and a product manager (PM).",
    )

# Initialize the group chat between the user and two LLM agents (PM and coder)
groupchat = autogen.GroupChat(agents=[user_proxy, coder, pm], messages=[], max_round=12)
manager = autogen.GroupChatManager(groupchat=groupchat, llm_config=llm_config)

# Begin the group chat with a message from the user
user_proxy.initiate_chat(
    manager,
    message="I want to design an app to make me one million dollars in one month. Yes, your heard that right.",
)
```

I get this output:
```text
$ python memgpt/autogen/examples/agent_groupchat.py

User_proxy (to chat_manager):

I want to design an app to make me one million dollars in one month. Yes, your heard that right.

--------------------------------------------------------------------------------
Product_manager (to chat_manager):

Creating an app that generates one million dollars in a month is certainly ambitious, and while I can't guarantee it will reach that goal, I can certainly help brainstorm some potentially profitable ideas!

Here are a few app ideas that with the right execution, marketing, and sales strategy could perform well in the market:

1. Mentor Matching App: A platform that connects experienced mentors with individuals looking to learn new skills or gain career advice. You'll generate revenue by taking a small percentage of the mentorship fees.

2. Virtual Reality Estate Planning: This concept uses virtual reality to help people plan their estates. It could provide a 3D visualization of how an individual's wealth and assets will be distributed.

3. Online Courses App: High-quality, frequently updated online courses and certifications in trending fields like cybersecurity, data science, AI, etc. Revenue could be through subscription fees and ads.

4. Personalized Diet & Meal Plan: An app that uses AI to recommend personalized diets, grocery lists, and meal plans based on users' dietary preferences, restrictions, allergy information, and health goals. Monetization can be done through subscription fees or partnerships with local grocery stores for referred purchases.

5. P2P Item Rental: People have stuff that they use rarely - why not rent it out to others who need it? From tools to designer clothes, this app can offer a short-term, cost-effective solution.

6. Virtual Interior Design: Utilize augmented reality to allow users to visualize potential interior design options in their exact space. You can earn by offering a premium plan or partnerships with furniture and home decor companies.

7. Mental Health Social Network: An app focused on mental health and wellbeing, offering peer support, expert advice, and resources. Monetization can be done through in-app purchases, premium membership, or ads run by certified wellness advertisers.

8. Smart City App: An app that aggregates all city services into a user-friendly platform. Parking, transit, city events, municipal services, etc. The app would generate revenue through ads and potentially city funding.

Remember, all great products start with an audience's niche needs then scale up. It's essential to validate the market demand before starting app development. Solid execution, user experience, marketing, and retention strategies are just as important as the original idea.

--------------------------------------------------------------------------------
MemGPT_coder (to chat_manager):

💭 Highly ambitious request. Validate feasibility with gentle humor and pepper it with optimism. Set realistic expectations while also offering a collaborative approach. Time to impress.
🤖 Well, as the first engineer at Uber, I love your 'go big or go home' attitude! Making millions in a month is hopeful but not impossible. We just need a groundbreaking idea and some solid groundwork. What's your app idea?
⚡🟢 [function] Success: None
💭 Assessing user response... Wow, this user has a plethora of ambitious ideas. Time to give a detailed analysis on their options while maintaining an encouraging tone. Demonstrate ability to strategize and help with this undertaking. Don't forget a bit of humor for camaraderie. Let's dig into my startup experience.
🤖 Wow, with such a list, we might make a million a week, not just a month! Each idea has its own charm and potential profitability. It seems all of them require robust tech—a perfect job for a 10x engineer! We need to identify our target audience, their needs, and our standing against competitors. Let's study each idea, deduce market demand, work on the MVPs, and ultimately choose the most feasible one. I'm excited about this journey. We might just give Uber a run for its money!
⚡🟢 [function] Success: None

--------------------------------------------------------------------------------

>>>>>>>> USING AUTO REPLY...
User_proxy (to chat_manager):



--------------------------------------------------------------------------------

>>>>>>>> USING AUTO REPLY...
User_proxy (to chat_manager):



--------------------------------------------------------------------------------
Product_manager (to chat_manager):

Here are few innovative software product ideas:

1. "Intelligent Assistant for Healthcare": An AI-powered tool that not only provides online consultation using user's symptoms and medical history but also sets reminders for taking medicines, connects with nearby pharmacies for pill-refill, helps book doctor appointments and maintains a digital health history.

2. "Green Thumb": A mobile application that helps amateur gardeners and plant lovers. The app can provide advice on plant care, pest identification, and remedies, seasonal plants information, gardening tools and their uses etc. The application could use image recognition to help identify plants and pests.

3. "Quick Tutors": An online platform connecting tutors and students in real-time. Tutors can create profiles listing their expertise and students can find tutors for specific subjects or topics on-demand for quick doubt resolution, assignment help or exam preparation. 

4. "Event Planner Assistant": A software product that would take the details of your event (like occasion, number of guests, date, location etc), ask you for preferences in type of cuisine, decor, music etc and use all this data to provide the best plans, vendors, and prices for you to choose from. 

5. "AR Historical Tour": An application that employs augmented reality to give users a tour of historical landmarks, showing what they would have looked like in their prime. Users simply point their device's camera to the landmark and the screen would display the AR version of it. 

6. "Freelance for Seniors": A platform dedicated for people aged 60 and above. There are plenty of seniors with valuable skills and a lot of free time. This software platform would help them connect with small business owners or individuals in need of their skills. 

7. "Auto Dietician": An app that tracks user's food intake, physical activities, and suggest diet plans based on their health goals. It could also integrate with smart devices to collect fitness data. 

Remember, a great product idea also needs good marketing strategy, user-friendly design and strong customer support for it to be successful.

--------------------------------------------------------------------------------

>>>>>>>> USING AUTO REPLY...
User_proxy (to chat_manager):
```

**Comment by cpacker** [comment]
Made a patch for the AutoGen bug - it seems like in the API call we're overriding, `messages` is the full message history.

We were also iterating over all the messages, which seems wrong even is `messages` is trimmed to only include new messages, because it makes MemGPT think these are interactive messages vs a past chat history.

I modified the override function to instead:
1. strip `messages` down to new messages only
2. either concat all the new messages into one user message, and feed that, or
3. silently extend MemGPT agent's message list with `new_messages[:-1]`, then take a step with `new_messages[-1]`

IMO (3) makes more sense to me than (2). However, (2) is technically closer to how the ChatCompletion API is meant to be used (alternating user/assistant messages). With (3), if you have N non-MemGPT agents in the group chat, you'll get sections of the messages list with bunches of N consecutive 'user' messages. This is probably OOD for the gpt-4 model but seems like it's robust enough to handle it fine.

File: memgpt/autogen/memgpt_agent.py [hunk 1]
```diff
@@ -1,13 +1,22 @@
 from autogen.agentchat import ConversableAgent, Agent
 from ..agent import AgentAsync
 
-from .. import system
-from .. import constants
+# from .. import system
+# from .. import constants
```
Comment (vivi) [inline memgpt/autogen/memgpt_agent.py:5]: remove leftover comments

**Comment by vivi** [comment]
Example output w/ simplified find_new_messages
```text
User_proxy (to chat_manager):

I want to design an app to make me one million dollars in one month. Yes, your heard that right.

--------------------------------------------------------------------------------

Product_manager (to chat_manager):

App Idea: "Investor Elite"

The app would be an elite subscription-based platform where high-profile investors and beginners can converge. This would provide an opportunity for budding investors to learn from top-notch experts from various industries globally.

Features:
1. Live Streams: Experts conduct live streams, sharing key investment insights and strategies. Users can tune in to listen, learn, and ask questions right away.

2. Investment Trends : The app would use AI to follow global markets and financial news to show up-to-the-minute investment trends.

3. Expert Consultation: Create a 1-on-1 investor experts consulting option where users can book sessions for personalized advice.

4. Investment Simulator: This feature can provide a real-time trading experience to the users with virtual money, helping them implement what they have learned.

5. Community: A forum for users to ask questions, share their experiences, strategies, and network with other investors.

Monetization:
1. Subscription fee: You can introduce a monthly/annual subscription fee to gain access to specialist live streams, expert consultations, and community privileges.

2. Sponsorships: Sponsorships from corporates or investment firms can help you earn additional income.

3. In-app purchases: Users can buy special features like additional educational materials, premium investment calculators, and more.

Target your marketing towards affluent investors and wealth management companies. If you can get even 5,000 users to sign up for a $200 premium monthly subscription, you can make a million in a month.

Bear in mind, no idea is a guaranteed success. It will require a lot of work, a solid business plan, and a bit of luck to achieve such ambitious revenue goals. But this app concept leverages a high-value market, offers attractive features, and has a strong monetization plan.

--------------------------------------------------------------------------------
MemGPT_coder (to chat_manager):

Sounds like a grand idea, Product_manager, and it could be feasible from a technical standpoint. We would need a team of skilled developers to ensure we're creating a highly secure, scalable, and interactive platform. AI will play a crucial role in some of the suggested features like tracking global markets and financial news for real-time trends. I'd also recommend that we start with developing a minimum viable product (MVP) with core features and gradually add more based on the users' requirements and feedback.

--------------------------------------------------------------------------------

>>>>>>>> USING AUTO REPLY...
User_proxy (to chat_manager):



--------------------------------------------------------------------------------
Product_manager (to chat_manager):

That sounds like a plan! To summarize, the initial steps would be:
1. Conduct market research to understand our target audience better.
2. Gather a committed, capable team.
3. Start the design and development with a focus on security, scalability, and user-friendly interfaces.
4. Prioritize and implement core features based on user requirements, market need, and technical feasibility.
5. Test the MVP extensively and improve it based on user feedback and market response.
6. Plan for future features and improvements.
7. Work on a marketing strategy and launch.
8. Continual monitoring, updates, and user engagement activities.

If you're ready to proceed, we can start the detailed planning for each step right away. I can help you outline a comprehensive plan to move forward with this exciting project.

--------------------------------------------------------------------------------

>>>>>>>> USING AUTO REPLY...
User_proxy (to chat_manager):



--------------------------------------------------------------------------------
MemGPT_coder (to chat_manager):

What are your thoughts, User_proxy? Your input is crucial in shaping the direction of this project. Are there any specific areas you'd like us to elaborate on? Or do you have other ideas you'd like to explore?

--------------------------------------------------------------------------------

>>>>>>>> USING AUTO REPLY...
User_proxy (to chat_manager):



--------------------------------------------------------------------------------
Product_manager (to chat_manager):

I'm glad you're considering this idea. Please feel free to share any specific expectations or requirements you have for this project. This can include insights into prospective target audiences, desired features, budget considerations, timeline, or any other specifications. With these details, we can formulate a more tailored and comprehensive action plan for your Million-Dollar-App. Also, if you have any concerns or questions, do let us know!

--------------------------------------------------------------------------------

>>>>>>>> USING AUTO REPLY...
User_proxy (to chat_manager):



--------------------------------------------------------------------------------
MemGPT_coder (to chat_manager):

I understand this is a big decision, User_proxy. Take your time and remember, we're here to make this vision a reality. Whether it's about the features, the design, or even the color scheme, no detail is too small. We're all in this together.

--------------------------------------------------------------------------------

>>>>>>>> USING AUTO REPLY...
User_proxy (to chat_manager):



--------------------------------------------------------------------------------
```

File: memgpt/autogen/memgpt_agent.py [hunk 2]
```diff
@@ -20,74 +29,158 @@ def create_memgpt_autogen_agent_from_config(
...
+        while True:
+            (
+                new_messages,
+                heartbeat_request,
+                function_failed,
+                token_warning,
+            ) = await self.agent.step(
+                user_message, first_message=False, skip_verify=self.skip_verify
+            )
+            # ret.extend(new_messages)
```
Comment (vivi) [inline memgpt/autogen/memgpt_agent.py:157]: stray comment

**Review by vivi — APPROVED** [review]
😈

**Comment by vivi** [comment]
> - [ ] Add TeachableAgent example (from https://github.com/microsoft/autogen/blob/main/test/agentchat/chat_with_teachable_agent.py / https://github.com/microsoft/autogen/blob/main/notebook/agentchat_teachability.ipynb)

Let's put that into a separate PR

**Comment by cpacker** [comment]
😈

LGTM

**Pull request merged**